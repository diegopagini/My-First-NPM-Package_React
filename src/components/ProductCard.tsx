/** @format */
import { createContext } from 'react';

import { useProduct } from '../hooks/useProduct';
import {
  InitialValues,
  OnChangeArgs,
  Product,
  ProductCardHandlers,
  ProductContextProps,
} from '../interfaces/product.interfaces';
import styles from '../styles/styles.module.css';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export interface Props {
	product: Product;
	// children?: ReactElement | ReactElement[];
	children: (args: ProductCardHandlers) => JSX.Element;
	className?: string;
	style?: React.CSSProperties;
	onChange?: (args: OnChangeArgs) => void;
	value?: number;
	initialValues?: InitialValues;
}

export const ProductCard = ({
	children,
	product,
	className,
	style,
	onChange,
	value,
	initialValues,
}: Props) => {
	const { counter, increaseBy, maxCount, isMaxCountReached, reset } = useProduct({
		onChange,
		product,
		value,
		initialValues,
	});

	return (
		<Provider
			value={{
				counter,
				increaseBy,
				product,
				maxCount,
			}}>
			{/* To use more than one style class */}
			<div className={`${styles.productCard} ${className}`} style={style}>
				{children({
					count: counter,
					maxCount: initialValues?.maxCount,
					product,
					increaseBy,
					isMaxCountReached,
					reset,
				})}
			</div>
		</Provider>
	);
};
