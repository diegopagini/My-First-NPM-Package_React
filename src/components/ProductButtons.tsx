/** @format */
import { useCallback } from 'react';
import { useContext } from 'react';

import styles from '../styles/styles.module.css';
import { ProductContext } from './ProductCard';

export interface Props {
	className?: string;
	style?: React.CSSProperties;
}

export const ProductButtons = ({ className, style }: Props) => {
	const { increaseBy, counter, maxCount } = useContext(ProductContext);

	const isMaxReached = useCallback(() => !!maxCount && counter === maxCount, [counter, maxCount]);

	return (
		<div className={`${styles.buttonsContainer} ${className}`} style={style}>
			<button onClick={() => increaseBy(-1)} className={styles.buttonMinus}>
				-
			</button>
			<div className={styles.countLabel}>{counter}</div>
			<button
				className={`${styles.buttonAdd} ${isMaxReached() && styles.disabled}`}
				onClick={() => increaseBy(+1)}>
				+
			</button>
		</div>
	);
};
