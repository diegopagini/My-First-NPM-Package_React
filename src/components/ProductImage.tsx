/** @format */
import { useContext } from 'react';

import noImage from '../assets/no-image.jpg';
import styles from '../styles/styles.module.css';
import { ProductContext } from './ProductCard';

export interface Props {
	img?: string;
	className?: string;
	style?: React.CSSProperties;
}

export const ProductImage = ({ img, className, style }: Props) => {
	const { product } = useContext(ProductContext);
	let imgToShow: string;
	if (img) imgToShow = img;
	else if (product.img) imgToShow = product.img;
	else imgToShow = noImage;

	return (
		<img
			className={`${styles.productImg} ${className}`}
			src={imgToShow}
			alt='Product'
			style={style}
		/>
	);
};
