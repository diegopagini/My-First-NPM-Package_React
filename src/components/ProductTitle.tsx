/** @format */
import { useContext } from 'react';

import styles from '../styles/styles.module.css';
import { ProductContext } from './ProductCard';

export interface Props {
	title?: string;
	className?: string;
	style?: React.CSSProperties;
}

export const ProducTitle = ({ title, className, style }: Props) => {
	const { product } = useContext(ProductContext);

	return (
		<span className={`${styles.productDescription} ${className}`} style={style}>
			{/* If title is passed by props is going to use title, if not, the one from the product. */}
			{title ? title : product.title}
		</span>
	);
};
