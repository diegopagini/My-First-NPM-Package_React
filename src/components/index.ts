/** @format */
import { ProductCardHOCProps } from '../interfaces/product.interfaces';
import { ProductButtons } from './ProductButtons';
import { ProductCard as ProductCardHOC } from './ProductCard';
import { ProductImage } from './ProductImage';
import { ProducTitle } from './ProductTitle';

export const ProductCard: ProductCardHOCProps = Object.assign(ProductCardHOC, {
	Buttons: ProductButtons,
	Image: ProductImage,
	Title: ProducTitle,
});

export default ProductCard;
