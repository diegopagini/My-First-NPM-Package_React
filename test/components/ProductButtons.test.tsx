import React from 'react';
import renderer from 'react-test-renderer';

import { ProductButtons } from '../../src/components/ProductButtons';
import { ProductCard } from '../../src/components/ProductCard';
import { PRODUCT } from '../data/products';

describe('ProductButtons', () => {
  test('Debe de mostrar el componente correctamente', () => {
    const wrapper = renderer.create(<ProductButtons />);

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('Debe de mostrar el componente', () => {
    const wrapper = renderer.create(
      <ProductCard product={PRODUCT}>{() => <ProductButtons />}</ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
