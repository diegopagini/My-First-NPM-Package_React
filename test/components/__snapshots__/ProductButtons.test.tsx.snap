// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductButtons Debe de mostrar el componente 1`] = `
<div
  className="productCard undefined"
>
  <div
    className="buttonsContainer undefined"
  >
    <button
      className="buttonMinus"
      onClick={[Function]}
    >
      -
    </button>
    <div
      className="countLabel"
    >
      0
    </div>
    <button
      className="buttonAdd false"
      onClick={[Function]}
    >
      +
    </button>
  </div>
</div>
`;

exports[`ProductButtons Debe de mostrar el componente correctamente 1`] = `
<div
  className="buttonsContainer undefined"
>
  <button
    className="buttonMinus"
    onClick={[Function]}
  >
    -
  </button>
  <div
    className="countLabel"
  />
  <button
    className="buttonAdd false"
    onClick={[Function]}
  >
    +
  </button>
</div>
`;
